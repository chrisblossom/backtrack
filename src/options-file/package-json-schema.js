/* @flow */

import Joi from 'joi';

const disallowed = [
    'name',
    'description',
    'keywords',
    'private',
    'preferGlobal',

    'repository',
    'bugs',
    'homepage',

    /**
     * Possibly allow these
     *
     * If added and not found, exit and ask to install
     */
    'bundledDependencies',
    'bundleDependencies',
    'optionalDependencies',
    'peerDependencies',
    'devDependencies',
    'dependencies',
].reduce((acc, key) => {
    return {
        ...acc,
        [key]: Joi.any()
            .forbidden()
            .label(`package.json mangedKey: ${key}`),
    };
}, {});

const packageJsonSchema = Joi.array()
    .items(Joi.object(disallowed))
    .label('packageJson')
    .single(true)
    .options({
        allowUnknown: true,
    });

export { packageJsonSchema };
