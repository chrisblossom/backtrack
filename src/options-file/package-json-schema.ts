import Joi from '@hapi/joi';

const disallowed = [
	'name',
	'description',
	'keywords',
	'preferGlobal',

	'repository',
	'bugs',
	'homepage',

	/**
	 * Possibly allow these
	 *
	 * If added and not found, exit and ask to install
	 */
	'bundledDependencies',
	'bundleDependencies',
	'optionalDependencies',
	'peerDependencies',
	'devDependencies',
	'dependencies',
].reduce((acc, key) => {
	return {
		...acc,
		[key]: Joi.any()
			.forbidden()
			.label(`package.json managedKey: ${key}`),
	};
}, {});

const packageJsonSchema = Joi.array()
	.items(Joi.object(disallowed))
	.label('packageJson')
	.single(true)
	.options({
		allowUnknown: true,
	});

export { packageJsonSchema };
