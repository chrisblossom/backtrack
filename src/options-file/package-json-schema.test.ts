import Joi from '@hapi/joi';
import { packageJsonSchema } from './package-json-schema';

describe('packageJsonSchema', () => {
	test('is valid', () => {
		const managedKeys = {
			scripts: { dev: 'backtrack dev' },
		};

		const isValid = Joi.validate(managedKeys, packageJsonSchema);

		expect(isValid.error).toEqual(null);
	});

	test('is invalid', () => {
		const managedKeys = {
			keywords: 'backtrack name',
		};

		const isValid = Joi.validate(managedKeys, packageJsonSchema);

		expect(isValid.error).toMatchSnapshot();
	});

	test('is valid no managedKeys', () => {
		const managedKeys = {};

		const isValid = Joi.validate(managedKeys, packageJsonSchema);

		expect(isValid.error).toEqual(null);
	});

	test('is valid undefined managedKeys', () => {
		const isValid = Joi.validate(undefined, packageJsonSchema);

		expect(isValid.error).toEqual(null);
	});
});
