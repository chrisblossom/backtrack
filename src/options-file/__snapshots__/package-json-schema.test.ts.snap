// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packageJsonSchema is invalid 1`] = `[ValidationError: single value of "packageJson" fails because [child "package.json managedKey: keywords" fails because ["package.json managedKey: keywords" is not allowed]]]`;
